components/common/src/types/mod.rs

```diff
@@ -15,3 +15,24 @@
 mod listen_ctl_addr;
 
 pub use self::listen_ctl_addr::ListenCtlAddr;
+
+/// Bundles up information about the user and group that a supervised
+/// service should be run as. If the Supervisor itself is running with
+/// root-like permissions, then these will be for `SVC_USER` and
+/// `SVC_GROUP` for a service. If not, it will be for the user the
+/// Supervisor itself is running as.
+///
+/// On Windows, all but `username` will be `None`. On Linux,
+/// `username` and `groupname` may legitimately be `None`, but `uid`
+/// and `gid` should always be `Some`.
+#[derive(Debug, Default)]
+pub struct UserInfo {
+    /// Windows required, Linux optional
+    pub username: Option<String>,
+    /// Linux preferred
+    pub uid: Option<u32>,
+    /// Linux optional
+    pub groupname: Option<String>,
+    /// Linux preferred
+    pub gid: Option<u32>,
+}
```

Comment:
This is more of a musing, but have you looked into whether or not it'd be beneficial to have platform-specific implementations of this to tighten up the `Option`-ness?

On the one hand, it'd be nice for the compiler to be more aware of these documented contracts. On the other hand, it might make for some rather convoluted code.

🤔

Reply:
That's definitely worth looking into. I'll investigate after merging this since there are other benefits in this PR I don't want to hold up.

Reply:
Agreed 👍 I wouldn't want to hold this PR up, either.

Reply:
Investigated here: https://github.com/habitat-sh/habitat/pull/6328